### PR TITLE
[Feature] Support configurable policy for query node, Add user level schedule policy

### PR DESCRIPTION
**Pull request by chyezh**
/kind improvement

related issue: #23716
related PR: #23718

- pick from 2.2.0 and refactor on master.
- Support configurable policy for query node.
- Add new user level schedule policy in query node.
- Support schedulable query task

Main change as following.

- Add username into internalpb (search/query).
- Refactor the default scheduler into concurrent-safe scheduler for adding schedule strategy conveniently.
- Move metric collecting logic together.
- Query tasks join the scheduler now.

# Test grouping NQ result:
## Test setup

10 collection
768dim*100000 for every collection.
search param:
nq: 15
max_concurrency: 50
rate: 50

## master implementation
average grouping nq: 180
average latency of search: 1.4s

image
image
image

## implementation of this PR:

average grouping nq: 187
average latency of search: [1.28, 1.57]
image
image
image

## conclusion
The average grouping nq metric is greatly related to the system load.
I this that the current implement of fifo scheduler is equal to the master version.

### CODE REVIEW

**Comment by chyezh** [comment]
@yah01 @aoiasd  @weiliu1031 
I picked some changes about query node scheduling on 2.2.0, and there are some conflicts with your commit. Please review the changes when you have time.

Main change as following.
1. Add username into internalpb (search/query).
2. Refactor the default scheduler into concurrent-safe scheduler for adding schedule strategy conveniently.
3. Move metric collecting logic together.
4. Query tasks join the scheduler now.

Bugfix to confirm:
1. Original `MaxGroupNQ` is limit the total NQ in 2.2.x, but limit the search task count in master.
image
2. `PreExecute` fail may hang the task forever.
image
3.  Fix counter of `GetWaitingTaskTotalNQ`

File: internal/querynodev2/tasks/concurrent_safe_scheduler.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+func newConcurrentSafeScheduler(policy schedulePolicy) Scheduler {
+	maxReadConcurrency := paramtable.Get().QueryNodeCfg.MaxReadConcurrency.GetAsInt()
+	log.Info("query node use concurrent safe scheduler", zap.Int("max_concurrency", maxReadConcurrency))
+	return &concurrentSafeScheduler{
+		policy:           policy,
+		addChan:          make(chan addTaskReq),
```
Comment (yah01) [inline internal/querynodev2/tasks/concurrent_safe_scheduler.go:23]: this channel should be buffered as that provides more chances to merge tasks, an optimization could be like scheduler.go: 104

**Review by yah01 — COMMENTED** [review]
We need a test to measure how this works for merging

File: internal/querynodev2/tasks/concurrent_safe_scheduler.go (hunk 1 above)
Comment (chyezh) [reply internal/querynodev2/tasks/concurrent_safe_scheduler.go]: I have add buffer channel for new scheduler in new change,
There is the testing result:
test setup:
- 10 collection
- 768dim*100000 for every collection.
- search param:
  - nq: 15
  - max_concurrency: 50 
  - rate: 50

master implementation:
- average grouping nq: 180
- average latency of search: 1.4s

implementation of this PR:
- average grouping nq: 187
- average latency of search: [1.28, 1.57]

The average grouping nq metric is greatly related to the system load.
I this that the current implement of fifo scheduler is equal to the master version.

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

File: internal/querynodev2/tasks/concurrent_safe_scheduler.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	var task Task
+	for {
+		s.setupReadyLenMetric()
+
+		var execChan chan Task
+		task, execChan = s.setupExecListener(task)
+
+		select {
+		case <-ctx.Done():
+			return
```
Comment (yah01) [inline internal/querynodev2/tasks/concurrent_safe_scheduler.go:89]: add a log here so we could know the scheduler quited

File: internal/querynodev2/tasks/concurrent_safe_scheduler.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	}
+}
+
+// exec exec the ready task in background continuously.
+func (s *concurrentSafeScheduler) exec(ctx context.Context) {
+	log.Info("start execute loop")
+	for {
+		select {
+		case <-ctx.Done():
+			return
```
Comment (yah01) [inline internal/querynodev2/tasks/concurrent_safe_scheduler.go:166]: same as line 89

File: internal/querynodev2/tasks/fifo_policy.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	}
+}
+
+// fifoPolicy is a fifo policy with merge queue.
+type fifoPolicy struct {
+	queue *mergeTaskQueue
+}
+
+// Push add a new task into scheduler, an error will be returned if scheduler reaches some limit.
+func (p *fifoPolicy) Push(task Task) (int, error) {
```
Comment (yah01) [inline internal/querynodev2/tasks/fifo_policy.go:22]: As this never returns an error, maybe remove error from the return values?

File: internal/querynodev2/tasks/concurrent_safe_scheduler.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	}
+}
+
+type addTaskReq struct {
+	task Task
+	err  chan<- error
+}
+
+// concurrentSafeScheduler is a general concurrent safe scheduler implementation by wrapping a schedule policy.
+type concurrentSafeScheduler struct {
```
Comment (yah01) [inline internal/querynodev2/tasks/concurrent_safe_scheduler.go:41]: simply name it scheduler (or any nice name), the concurrent safe property is not important for this as it's not a data structure, and have to be always safe in concurrency env

File: pkg/util/paramtable/component_param.go [hunk 6]
```diff
@@ -1445,6 +1445,13 @@ type queryNodeConfig struct {
 
 	// loader
 	IoPoolSize ParamItem `refreshable:"false"`
+
+	// schedule task policy.
+	SchedulePolicyName                    ParamItem `refreshable:"false"`
+	SchedulePolicyMaxPendingTask          ParamItem `refreshable:"true"`
```
Comment (yah01) [inline pkg/util/paramtable/component_param.go:1451]: how this works diff to MaxReadConcurrency, should it be always not greater than MaxReadConcurrency?

File: internal/querynodev2/tasks/fifo_policy.go (hunk 4 above)
Comment (chyezh) [reply internal/querynodev2/tasks/fifo_policy.go:22]: `fifo` implements a `schedulePolicy` interface,
Another `user-task-polling` policy returns error if some limit reached.

File: pkg/util/paramtable/component_param.go (hunk 6 above)
Comment (chyezh) [reply pkg/util/paramtable/component_param.go:1451]: I think that `MaxReadConcurrency` is no comparison relationship with `SchedulePolicyMaxPendingTask`.

- `MaxReadConcurrency` limits the max concurrent running tasks.
- `SchedulePolicyMaxPendingTask` limits the max pending tasks. query scheduler will reject new incoming task if the pending number of task exceeds this parameter. If the pending task queue is too long, the scheduler will cost much time to handle merging logic. It is just a protective logic that will not trigger in normal usage mode.

**Comment by chyezh** [comment]
I have add buffer channel for new scheduler in new change,
There is the testing result:
test setup:

10 collection
768dim*100000 for every collection.
search param:
nq: 15
max_concurrency: 50
rate: 50
master implementation:

average grouping nq: 180
average latency of search: 1.4s
implementation of this PR:

average grouping nq: 187
average latency of search: [1.28, 1.57]
The average grouping nq metric is greatly related to the system load.
I this that the current implement of fifo scheduler is equal to the master version.

**Comment by chyezh** [comment]
rebase from master

File: internal/querynodev2/tasks/queues.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+// tryMerge try to a new task to any task in queue.
+func (q *mergeTaskQueue) tryMerge(task MergeTask, maxNQ int64) bool {
+	nqRest := maxNQ - task.NQ()
+	// No need to perform any merge if task.nq is greater than maxNQ.
+	if nqRest < 0 {
+		return false
+	}
+	for i := 0; i < q.len(); i++ {
```
Comment (yah01) [inline internal/querynodev2/tasks/queues.go:69]: reverse iteration takes higher probability to return early

Comment (chyezh) [reply internal/querynodev2/tasks/queues.go:69]: - Reverse iteration do not make great improvement on my last test.
- Task of same segment may locate at random positions in the queue.

**Comment by chyezh** [comment]
rerun ut

File: internal/querynodev2/tasks/concurrent_safe_scheduler.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+			log.Warn("unexpected quit of schedule loop")
+			return
+		case req := <-s.receiveChan:
+			// Receive add operation request and return the process result.
+			// And consume recv chan as much as possible.
+			s.consumeRecvChan(req, maxReceiveChanBatchConsumeNum)
+		case execChan <- task:
+			// Task sent, drop the ownership of sent task.
+			// And produce new task into execChan as much as possible.
+			task = s.produceExecChan()
```
Comment (congqixia) [inline internal/querynodev2/tasks/concurrent_safe_scheduler.go:98]: If a task is return, it may be overwritten by line 85 and cause task go orphan

Comment (chyezh) [reply internal/querynodev2/tasks/concurrent_safe_scheduler.go:98]: setupExecListener (function at line 85) only overwrite task if the task == nil.
So task waiting to send execChan will never go orphan.

**Review by congqixia — APPROVED** [review]
/kind feature
/lgtm

**Pull request merged**